Add invariant that all symbolic shapes must be bound in graph

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99089

Previously, we had a problem when partitioning forward-backward dynamic graphs, which is that we could end up with a backward graph that mentions a symbol in an input tensor (e.g., `f32[s0 + s1]`), but without this symbol being otherwise bound elsewhere. When this happens, we have no way of actually deriving the values of `s0` and `s1`. Our fix for this in https://github.com/pytorch/pytorch/pull/93059 was to just retrace the graph, so that s0 + s1 got allocated a new symbol s2 and everything was happy. However, this strategy had other problems, namely (1) we lost all information from the previous ShapeEnv, including guards and (2) we end up allocating a LOT of fresh new symbols in backwards.

With this change, we preserve the same ShapeEnv between forward and backwards. How do we do this? We simply require that every symbol which may be present inside tensors, ALSO be a plain SymInt input to the graph. This invariant is enforced by Dynamo. Once we have done this, we can straightforwardly modify the partitioner to preserve these SymInt as saved for backwards, if they are needed in the backwards graph to preserve the invariant as well.

This apparently breaks yolov3, but since everything else is OK I'm merging this as obviously good and investigating later.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

ezyang committed:
Add invariant that all symbolic shapes must be bound in graph

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -344,6 +338,48 @@ def restore_graphstate(self, state: OutputGraphState):
+        if self.export:
+            return
+
+        if arg.fake_tensor is None:
+            return
```
When does this happen?

ezyang replied:
It doesn't, can assert.

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -814,18 +865,53 @@ def remove_unused_graphargs(self) -> None:
+        def placeholder_binds_symbol(node):
+            arg = node.meta["grapharg"]
+            example = arg.example
+            if isinstance(example, torch.SymInt) and isinstance(
+                example.node.expr, sympy.Symbol
+            ):
```
Can you explain this condition?

ezyang replied:
We're looking for an input which is a SymInt and is a symbol and not some derived expression. I can have SymInt that expresses, e.g., s0 + s1 (or s0 // 2) and this doesn't "bind" the symbol, because you cannot infer s0/s1 from just the expression.

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -814,18 +865,53 @@ def remove_unused_graphargs(self) -> None:
-                self.remove_node(node)
-                self.real_value_cache.pop(node, None)
+        def placeholder_binds_symbol(node):
+            arg = node.meta["grapharg"]
```
nit: you do `["grapharg"]` in enough places across a few files to warrant giving it a const imo

ezyang replied:
We don't const "val" or "example_input" though?

voznesenskym replied:
No no, I mean, the string! Like, just do something like

META_GRAPHARG_KEY = "grapharg"

nit anyway.

ezyang replied:
I understood you. We don't have a key like this for "example_input". Are you saying we should have a key for that too?

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -814,18 +865,53 @@ def remove_unused_graphargs(self) -> None:
+        used_symbols = set()
+        recheck_placeholders = []
+        for node in self.placeholders:
+            binds_symbol = placeholder_binds_symbol(node) is not None
+            # Don't delete symbol bindings yet
+            if binds_symbol:
+                if not node.users:
+                    recheck_placeholders.append(node)
```
When is this the case? why do we need to recheck? is this more like, unused_bound_symbols?

ezyang replied:
They're all the placeholders that we need to check again after we've found out what symbols we need to bind. They are not unused bound symbols because they might be implicitly "used" by other tensors which refer to their symbol.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -170,6 +170,41 @@ def is_concrete_int(a: Union[int, SymInt]):
+        return set()
+    elif isinstance(val, torch.Tensor):
+        r = set()
+        for s in val.size():
+            r |= free_symbols(s)
+        for s in val.stride():
+            r |= free_symbols(s)
+        r |= free_symbols(val.storage_offset())
+        return r
```
Do you care if these are not symbolic? I know you handle int up there, but is there ever a case you want this to work on just a regular tensor of regular int sizes/strides?

ezyang replied:
I mean, if it's a regular tensor, free_symbols will always return empty set.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -170,6 +170,41 @@ def is_concrete_int(a: Union[int, SymInt]):
+    else:
+        raise AssertionError(f"cannot compute free_symbols of {val}")
+
+# WARNING: Don't use this on Dynamo produced graphs, they don't have meta
+# setup!
+def is_symbol_binding_fx_node(node) -> Optional[sympy.Symbol]:
```
Yes, this is a refactor I was thinking about. One good start would be to change Dynamo to use "val" instead of "example_input". Another would be top stop deleting example_input after we're done compiling.

ezyang committed:
Update on "Add invariant that all symbolic shapes must be bound in graph"


Previously, we had a problem when partitioning forward-backward dynamic graphs, which is that we could end up with a backward graph that mentions a symbol in an input tensor (e.g., `f32[s0 + s1]`), but without this symbol being otherwise bound elsewhere. When this happens, we have no way of actually deriving the values of `s0` and `s1`. Our fix for this in https://github.com/pytorch/pytorch/pull/93059 was to just retrace the graph, so that s0 + s1 got allocated a new symbol s2 and everything was happy. However, this strategy had other problems, namely (1) we lost all information from the previous ShapeEnv, including guards and (2) we end up allocating a LOT of fresh new symbols in backwards.

With this change, we preserve the same ShapeEnv between forward and backwards. How do we do this? We simply require that every symbol which may be present inside tensors, ALSO be a plain SymInt input to the graph. This invariant is enforced by Dynamo. Once we have done this, we can straightforwardly modify the partitioner to

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Add invariant that all symbolic shapes must be bound in graph"


Previously, we had a problem when partitioning forward-backward dynamic graphs, which is that we could end up with a backward graph that mentions a symbol in an input tensor (e.g., `f32[s0 + s1]`), but without this symbol being otherwise bound elsewhere. When this happens, we have no way of actually deriving the values of `s0` and `s1`. Our fix for this in https://github.com/pytorch/pytorch/pull/93059 was to just retrace the graph, so that s0 + s1 got allocated a new symbol s2 and everything was happy. However, this strategy had other problems, namely (1) we lost all information from the previous ShapeEnv, including guards and (2) we end up allocating a LOT of fresh new symbols in backwards.

With this change, we preserve the same ShapeEnv between forward and backwards. How do we do this? We simply require that every symbol which may be present inside tensors, ALSO be a plain SymInt input to the graph. This invariant is enforced by Dynamo. Once we have done this, we can straightforwardly modify the partitioner to

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Add invariant that all symbolic shapes must be bound in graph"


Previously, we had a problem when partitioning forward-backward dynamic graphs, which is that we could end up with a backward graph that mentions a symbol in an input tensor (e.g., `f32[s0 + s1]`), but without this symbol being otherwise bound elsewhere. When this happens, we have no way of actually deriving the values of `s0` and `s1`. Our fix for this in https://github.com/pytorch/pytorch/pull/93059 was to just retrace the graph, so that s0 + s1 got allocated a new symbol s2 and everything was happy. However, this strategy had other problems, namely (1) we lost all information from the previous ShapeEnv, including guards and (2) we end up allocating a LOT of fresh new symbols in backwards.

With this change, we preserve the same ShapeEnv between forward and backwards. How do we do this? We simply require that every symbol which may be present inside tensors, ALSO be a plain SymInt input to the graph. This invariant is enforced by Dynamo. Once we have done this, we can straightforwardly modify the partitioner to

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Add invariant that all symbolic shapes must be bound in graph"


Previously, we had a problem when partitioning forward-backward dynamic graphs, which is that we could end up with a backward graph that mentions a symbol in an input tensor (e.g., `f32[s0 + s1]`), but without this symbol being otherwise bound elsewhere. When this happens, we have no way of actually deriving the values of `s0` and `s1`. Our fix for this in https://github.com/pytorch/pytorch/pull/93059 was to just retrace the graph, so that s0 + s1 got allocated a new symbol s2 and everything was happy. However, this strategy had other problems, namely (1) we lost all information from the previous ShapeEnv, including guards and (2) we end up allocating a LOT of fresh new symbols in backwards.

With this change, we preserve the same ShapeEnv between forward and backwards. How do we do this? We simply require that every symbol which may be present inside tensors, ALSO be a plain SymInt input to the graph. This invariant is enforced by Dynamo. Once we have done this, we can straightforwardly modify the partitioner to

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh commented on test/dynamo/test_dynamic_shapes.py:
```diff
@@ -31,8 +31,6 @@
         # Could not infer dtype of torch._C.SymIntNode
         "test_convert_boxes_to_pooler_format",
-        # TODO: look into this: FakeTensor error.
-        "test_gan_repro_trying_to_backward_through_the_graph_a_second_time",
```
Thanks :)

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2590,24 +2590,11 @@ def backward(ctx, *flat_args):
+                    context = disable_autocast_manager if disable_amp else nullcontext
+                    with context(), track_graph_compiling(aot_config, "backward"):
+                        CompiledFunction.compiled_bw = aot_config.bw_compiler(
+                            bw_module, all_args
```
This should also fix https://github.com/pytorch/pytorch/issues/98416 ?

ezyang replied:
Yep!

bdhirsh commented on torch/_dynamo/output_graph.py:
```diff
@@ -334,6 +348,47 @@ def restore_graphstate(self, state: OutputGraphState):
+        # having to pull it out implicitly from tensors.
+
+        if self.export:
+            return
```
oh maybe the answer is no, because we don't do any partitioning - so we don't have to worry about "I have a separate bw graph that refers to symbols that don't exist because they were created in the forward graph"

ezyang replied:
If we did want to do dynamic + export + partitioning we would be in some trouble. But we can cross that road when we get there.

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -170,6 +170,41 @@ def is_concrete_int(a: Union[int, SymInt]):
+        node.op == "placeholder" and
+        "val" in node.meta and
+        isinstance(node.meta["val"], torch.SymInt) and
+        isinstance(node.meta["val"].node.expr, sympy.Symbol)
```
Checking my understanding: is it possible to have a SymInt-bound placeholder node, who's `node.meta["val"].node.expr` is something other than ` sympy.Symbol`? If so, when? (and if not maybe we should assert against it).

ezyang replied:
Yes, is possible. While it is true that we only ever add placeholder nodes for fresh symbols, a symbol can get refined as we do more operations, and then it turns out that it's a composite. For example:

```
def f(s0, s1):
    assert s0 == s1 * 2
```

The s0 will become s1 * 2 and no longer a sympy.Symbol.

bdhirsh replied:
Ah, right. Thanks for the example!

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -120,6 +125,46 @@ def _extract_fwd_bwd_modules(joint_module: fx.GraphModule, saved_values, saved_s
+            if s not in symbol_bindings:
+                continue
+            saved_sym_nodes_binding.append(symbol_bindings[s])
+        saved_symbols |= new_symbols
```
I agree that this is strictly better, but mostly just confirming: Instead of doing all of this pruning, could you just take every SymInt GraphArg from the forward, and force it to be saved for the backward?

I can buy the moral argument that we shouldn't bother saving things for backwards unnecessarily if we don't need them, although it probably wouldn't impact perf (since we're just saving sympy expressions / ints)

ezyang replied:
Yes, it is permissible not to do pruning.  Less things saved for backwards actually does marginally improve perf, because stashing things on autograd Function has to be done at runtime. Individually it doesn't matter, but I've seen some graphs with hundreds of symbols... seems suspicious.

ezyang commented:
I'm going to skip yolov3 for now and fix in follow up

ezyang committed:
Update on "Add invariant that all symbolic shapes must be bound in graph"


Previously, we had a problem when partitioning forward-backward dynamic graphs, which is that we could end up with a backward graph that mentions a symbol in an input tensor (e.g., `f32[s0 + s1]`), but without this symbol being otherwise bound elsewhere. When this happens, we have no way of actually deriving the values of `s0` and `s1`. Our fix for this in https://github.com/pytorch/pytorch/pull/93059 was to just retrace the graph, so that s0 + s1 got allocated a new symbol s2 and everything was happy. However, this strategy had other problems, namely (1) we lost all information from the previous ShapeEnv, including guards and (2) we end up allocating a LOT of fresh new symbols in backwards.

With this change, we preserve the same ShapeEnv between forward and backwards. How do we do this? We simply require that every symbol which may be present inside tensors, ALSO be a plain SymInt input to the graph. This invariant is enforced by Dynamo. Once we have done this, we can straightforwardly modify the partitioner to

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Add invariant that all symbolic shapes must be bound in graph"


Previously, we had a problem when partitioning forward-backward dynamic graphs, which is that we could end up with a backward graph that mentions a symbol in an input tensor (e.g., `f32[s0 + s1]`), but without this symbol being otherwise bound elsewhere. When this happens, we have no way of actually deriving the values of `s0` and `s1`. Our fix for this in https://github.com/pytorch/pytorch/pull/93059 was to just retrace the graph, so that s0 + s1 got allocated a new symbol s2 and everything was happy. However, this strategy had other problems, namely (1) we lost all information from the previous ShapeEnv, including guards and (2) we end up allocating a LOT of fresh new symbols in backwards.

With this change, we preserve the same ShapeEnv between forward and backwards. How do we do this? We simply require that every symbol which may be present inside tensors, ALSO be a plain SymInt input to the graph. This invariant is enforced by Dynamo. Once we have done this, we can straightforwardly modify the partitioner to preserve these SymInt as saved for backwards, if they are needed in the backwards graph to preserve the invariant as well.

This apparently breaks yolov3, but since everything else is OK I'm merging this as obviously good and investigating later.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
